crates/sui-indexer/migrations/pg/2024-09-12-213234_watermarks/up.sql

```diff
@@ -0,0 +1,22 @@
+CREATE TABLE watermarks
+(
+    -- The table name of group of tables governed by this watermark, i.e `epochs`, `checkpoints`,
+    -- `transactions`.
+	entity                      TEXT          NOT NULL,
+    -- Upper bound epoch range to enable per-entity epoch-level retention policy.
+    epoch_hi                    BIGINT        NOT NULL,
+    -- Lower bound epoch range to enable per-entity epoch-level retention policy.
+    epoch_lo                    BIGINT        NOT NULL,
+    -- The inclusive high watermark that the committer advances.
+	hi                          BIGINT        NOT NULL,
+    -- The inclusive low watermark that the pruner advances. Data before this watermark is
+    -- considered pruned.
+	lo                          BIGINT        NOT NULL,
+    -- Pruner sets this, and uses this column to determine whether to prune or wait long enough that
+    -- all in-flight reads complete or timeout before it acts on an updated watermark.
+	timestamp_ms                BIGINT        NOT NULL,
+    -- Pruner updates this, and uses this when recovering from a crash to determine where to
+    -- continue pruning.
+    pruned_lo                   BIGINT,
```

Comment:
Could you add explaination how is this different from `lo`?

Reply:
My understanding is that you want a watermark that is more stable for the reader (lo) and a mark that tracks true progress of the pruner. This means the lo mark for the reader wont change as frequently (read constantly) and can provide a bit of consistency for readers. So i'd expect the flow would be:

1. pruner decides which range to prune
2. update the lo mark to the top of that range (pruned_lo should be at the bottom of that range)
3. reader immediately acts as if everything has been pruned
4. pruner can start pruning from the bottom of the range to the top, updating the pruned_lo mark as needed

Reply:
Yes that's right, the `pruned_lo` and `timestamp_ms` are mostly there in the rare occasion where the pruner crashes and is restarted. Once it gets back up, rather than starting the pruning task immediately, and starting from the absolute beginning, it can consult `timestamp_ms` to determine whether it ought to prune, and `pruned_lo` for its actual progress, and resume from there.

Meanwhile, readers use `lo` and ignore everything below it